import "galaxy.proto";

package baidu.galaxy;

option cc_generic_services = true;

enum ProcessStatus {
    kProcessRunning = 0;
    kProcessTerminate = 1;
}

message ProcessInfoCheckpoint {
    repeated ProcessInfo process_infos = 1;
}

message ProcessInfo {
    optional string key = 1;
    optional int32 pid = 2; 
    optional ProcessStatus status = 3;
    optional int32 exit_code = 4;
}

message GetProcessStatusRequest {
    optional string key = 1;
}

message GetProcessStatusResponse {
    optional ProcessInfo process = 1;
    optional Status status = 2;
}

message ExecuteRequest {
    optional string key = 1; 
    optional string commands = 2;
    repeated string envs = 3;
    optional string path = 4;
    optional string cgroup_path = 5;
}

message ExecuteResponse {
    optional string key  = 1;
    optional int32 pid = 2;
    optional Status status = 3;
}

message GetPodStatusRequest {
    optional string key = 1;
}

message GetPodStatusResponse {
    optional Status status = 1;
    optional PodStatus pod_status = 2;
}

message CreatePodRequest {
    optional string podid = 1;
    optional PodDescriptor pod = 2;
}

message CreatePodResponse {
    optional Status status = 1;
}

message DeletePodRequest {
    optional string podid = 1;
}

message DeletePodResponse {
    optional Status status = 1;
}

message UpdatePodRequest {
    optional string podid = 1;
    optional Resource usage = 2;
}

message UpdatePodResponse {
    optional Status status = 1;
}

service Initd {
    // deprecated
    rpc GetProcessStatus(GetProcessStatusRequest) returns (GetProcessStatusResponse);
    // deprecated
    rpc Execute(ExecuteRequest) returns (ExecuteResponse);

    rpc GetPodStatus(GetPodStatusRequest) returns (GetPodStatusResponse);
    rpc CreatePod(CreatePodRequest) returns (CreatePodResponse);
    rpc DeletePod(DeletePodRequest) returns (DeletePodResponse);
    rpc UpdatePod(UpdatePodRequest) returns (UpdatePodResponse);
}
